### DISCUSSION TITLE
도메인 간, 쿠키가 공유되지 않을 때, 어떻게 해야 할까?

### DISCUSSION THREAD

**Discussion by devjun10** [dailyge/dailyge-server#105] [Information]
## 🔗 연관 PR

- [[DLG-172] 서브 도메인에도 쿠키가 공유될 수 있도록 한다.](https://github.com/dailyge/dailyge-server/pull/104)

<br/><br/>

## ❓ 작성 이유

프로젝트에서 서브 도메인을 사용하고 있기 때문에, 쿠키 설정을 추가하며 관련 내용을 공유하기 위해 글을 작성하게 되었습니다.

> [In the Domain Name System hierarchy, a subdomain is a domain that is a part of another (main) domain. For example, if a domain offered an online store as part of their website example.com, it might use the subdomain shop.example.com .](https://en.wikipedia.org/wiki/Subdomain)

<br/><br/>

## 📝 내용

현재 클라이언트 서버는 www.dailyge.com, tasks.dailyge.com을 사용하고 있으며, 서버는 api.dailyge.com을 사용하고 있습니다. 이 외에도 개발 서버에도 서브 도메인을 사용하고 있고요. 서브 도메인에는 설정이 없을 시, 쿠키가 공유되지 않을 수 있기 때문에 하위 도메인에 쿠키를 전달 할 수 있도록 설정을 추가해야 합니다. 설정 자체는 간단한데요, 쿠키의 domain 필드에 최상위 도메인을 명시해주면 됩니다. 

```java
@PresentationLayer
@RequestMapping("/api")
@RequiredArgsConstructor
public class LoginApi {

    ......

    @GetMapping(path = "/oauth2")
    public ApiResponse<OAuthLoginResponse> login(@RequestParam("code") final String code) {
        final DailygeToken dailygeToken = userFacade.login(code);
        final HttpHeaders headers = new HttpHeaders();
        final ResponseCookie refreshTokenCookie = ResponseCookie.from("Refresh-Token", dailygeToken.refreshToken())
            .domain(".dailyge.com")    // 최상위 도메인
            .path("/")
            .httpOnly(true)
            .secure(true)
            .build();
        headers.add(SET_COOKIE, refreshTokenCookie.toString());
        final OAuthLoginResponse payload = new OAuthLoginResponse(dailygeToken.accessToken());
        return ApiResponse.from(OK, headers, payload);
    }

    ......

}
```

<br/><br/><br/><br/>

즉, 현재 저희 최상위 도메인은 dailyge.com 이기 때문에 해당 설정을 추가했습니다. 

![image](https://img1.daumcdn.net/thumb/R1280x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2FdtQqor%2FbtsJc37qg3G%2FT87TbKrayh8AGMLlszAsI0%2Fimg.png)

<br/><br/><br/><br/>

## ✏️ 추가 학습

이 외에도 Cross Site 쿠키 전송, 쿠키에도 Lax, Strict, None 등 다양한 설정이 존재하기 때문에 관심 있으며 한 번 학습해보면 좋을 것 같습니다. 

- [SameSite cookies explained | Articles](https://web.dev/articles/samesite-cookies-explained)
- [Understanding SameSite cookies](https://andrewlock.net/understanding-samesite-cookies/)

<br/><br/><br/><br/>

## 📚 참조

- [RFC-6265](https://datatracker.ietf.org/doc/html/rfc6265)
- [Share cookies between subdomain and domain](https://stackoverflow.com/questions/18492576/share-cookies-between-subdomain-and-domain)
- [How to share the cookies between subdomains](https://abp.io/community/articles/how-to-share-the-cookies-between-subdomains-jfrzggc2)
- [Cross-Site Cookies - Understanding the Domain and ...](https://docs.descope.com/knowledgebase/security/crosssitecookies/)

**Comment by beatmeJY** [comment]
저는 서브 도메인은 생각을 해보지 못했었는데, 이렇게 최상위 도메인과 서브 도메인으로 쿠키를 공유하는 방법도 있군요! 코드부터 인프라 사용법까지도 좋은 글 작성해주셔서 감사합니다!👍🏻
